### PR TITLE
Segment form validation with meaningful error responses

### PR DESCRIPTION
**Pull request by kiambogo**

### CODE REVIEW

**Comment by kiambogo** [comment]
Example:

 ```
PATCH /segments/cc1992e6-1866-4e02-b317-0317b64a892c/rating/traffic/44 HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 0
Host: localhost:9000
User-Agent: HTTPie/0.9.2



HTTP/1.1 400 Bad Request
Content-Length: 31
Content-Type: application/json
Date: Mon, 10 Apr 2017 23:41:29 GMT

[
    {
        "error": "Rating > 5 or < 0"
    }
]

```

or 
```
POST /segments HTTP/1.1
Accept: application/json
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 158
Content-Type: application/json
Host: localhost:9000
User-Agent: HTTPie/0.9.2

{
    "description": "Ray's test segment",
    "name": "Segment 1",
    "points": [
        {
            "lat": 43.252416,
            "lng": -870.29559394
        },
        {
            "lat": 43.265773106,
            "lng": -80.2188109731
        }
    ]
}

HTTP/1.1 400 Bad Request
Content-Length: 58
Content-Type: application/json
Date: Mon, 10 Apr 2017 23:32:59 GMT

[
    {
        "error": "Point(43.252416,-870.29559394) out of bounds"
    }
]

```

**Comment by wittjosiah** [comment]
What's out of bounds? Like outside of the current map in the db?

File: api/app/controllers/Segments.scala [hunk 1]
```diff
@@ -39,15 +42,24 @@ class Segments @Inject() (segmentsDao: SegmentsDao, authLoggingAction:AuthLoggin
 
   def post() = AuthLoggingAction.async(parse.json[SegmentForm]) { implicit request =>
     val segForm = request.body
-    segmentsDao.upsert(segForm).map(s => Ok(Json.toJson(s)))
+    FormValidator.validateSegmentForm(segForm) match {
+      case Nil => println("No errors"); segmentsDao.upsert(segForm).map(s => Ok(Json.toJson(s)))
```
Comment (wittjosiah) [inline api/app/controllers/Segments.scala]: do we need to log "No errors"?

File: api/app/models/FormValidator.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package models
+
+import com.trifectalabs.roadquality.v0.models._
+
+object FormValidator {
+  def validateSegmentForm(segForm: SegmentForm): Seq[FormError] = {
+    { segForm.points.map { p =>
+        if (p.lat > 180 || p.lat < -180 || p.lng > 180 || p.lng < -180) Some(FormError(s"$p out of bounds"))
```
Comment (wittjosiah) [inline api/app/models/FormValidator.scala:8]: Ah I see, just check to make sure we're on earth 🙂

**Review by wittjosiah — APPROVED** [review]

File: api/app/controllers/Segments.scala (hunk 1 above)
Comment (kiambogo) [reply api/app/controllers/Segments.scala]: Nope, left that in when debugging. Thanks :)

File: api/app/models/FormValidator.scala (hunk 2 above)
Comment (kiambogo) [reply api/app/models/FormValidator.scala:8]: Yes - this is just a sanity check to ensure that the points are valid (on earth). As we roll-out city/country support we can think through an implementation to ensure that segments are only being created in the correct locations though.

**Pull request merged**